### ISSUE TITLE
Add support for vertex maps

### ISSUE THREAD

**Issue by douglaskastle** [douglaskastle/blender-import-lwo#14]
As requested by @Donald-Drennan

Found in this post on blender artists:

https://blenderartists.org/t/where-did-lightwave-import-export-go-in-2-57/506000/20?u=mavek

**Comment by douglaskastle** [comment]
@Donald-Drennan

Can you tell me what you are looking for in this mesh. You call it morph and it had morph keys.  The morph keys are being assigned in the mesh assignment.

**Comment by Donald-Drennan** [comment]
Yes, morph keys (morph targets). They should translate to Blender shape
keys. Can that be done?

Then there are vertex maps, and vertex paint maps.

Thanks again for your help on this!

**Comment by douglaskastle** [comment]
Can you import with the current addon, it seems to be pulling in both Shape Keys and Vertex colors:

![morph0](https://user-images.githubusercontent.com/79840/80894658-c4be9080-8cd4-11ea-9715-28d8c8f376ff.png)

Maybe you could explain a bit about what a vertex map is in LW and what it should be mapped to in Blender. Ditto for Vertex paint maps.

**Comment by Donald-Drennan** [comment]
That's great news! I haven't been able to install the add-on. Something I
was doing wrong? I'm not sure what the problem was. I will try your latest
version to see if it works.

Vertex maps equivalent to vertex maps in Blender. Used for selection sets
and other things. The same with vertex paint. Very similar to weight maps,
but for surface color. Each vertex is assigned a color. It's for rough
coloring of objects.

Hope that helps. For some reason I thought you were a Lightwave user too.
Thanks!

**Comment by douglaskastle** [comment]
I dabbled in LW, but never used any thing fancy like a morph or a vertex paint.  I am unsure if the current LWO parser is even pulling out the vertex paint information to translate it. But UV maps are there as are vertex colors, based on the image above.  I guess i need to know where I am coming from to know where I am going. So an example of the thing in LW that is not in blender would be helpful.  A whole mesh while a step it still too amorphous.

**Comment by Donald-Drennan** [comment]
Oh, vertex color and vertex paint would be the same thing. It's a method to
color an object. I never used that feature myself.  Morphs on the other
hand are very useful for animation. I used them to make facial expressions
in my mouse character, but they didn't transfer with obj format.

Thanks again?


Just let me know what you need.

**Comment by douglaskastle** [comment]
If you could resolve your addon install issue and check to see if the current implementation does what you need that would be very helpful.  

From my view point it is doing both of these things to some degree.  If they are not satisfactory then that is something else. 

It is unclear to me what work needs, if any, needs to be done on this issue.

**Comment by joewilliamsen** [comment]
@Donald:  Make sure you're not doing what I was doing when I had the same issue as described here: https://github.com/nangtani/blender-import-lwo/issues/12

FWIW, I was seeing morph targets (shape keys) being imported with the LWO model without a problem with the current and previous versions of the addon.

**Comment by Donald-Drennan** [comment]
Douglas,

I got the latest version of the add-on installed and tested it.

Morphs/shape keys works! I got my mouse character loaded in and all the
shape keys are there. Now I have to see if I can transfer the new UV map
and materials I made working on a version imported via OBJ file.

There seems to be something wrong with UV maps though. Lightwave vs.
Blender.
[image: Screenshot 2020-05-03 17.02.44.jpg]

[image: Screenshot 2020-05-03 17.04.37.jpg]
Vertex paint didn't exactly work. It imported them, but they are blank, no
color information.
Selection sets didn't work either, but this isn't very important to me.

Weight maps work fine.

Thanks so much for your help with this! This is so much better than it was.
I think I'll be able to import more of my models now without so much
headache.


Donald L. Drennan

**Comment by douglaskastle** [comment]
@Donald-Drennan The materials for cycles in the addon are not completed.  This is actually an issue that @joewilliamsen has reported and is the purpose of issue #9 to get that working.

Until then one can build the materials them selves, which seems to be the standard MO when importing meshes elsewhere.

I imagine the Vertex paint having no color is related to this.

However morph/shape are working, weight maps are working, thats great 

Let me check a few things, but I think this issue can be closed.

**Comment by Donald-Drennan** [comment]
Douglas,

Yes, I knew materials weren't really working. It does create materials, but
the images aren't connected in nodes. But, I can see how Lightwave's
surfaces aren't like Blender's materials.

Can you tell what's causing the UV map problem? Points in the UVs being
connected where they shouldn't be? It makes the UV maps kind of useless.

Thanks!

**Comment by douglaskastle** [comment]
I am looking into this now and I think the problem and the solution has to do with meshes that have more than one UV.  

If you use an image that is assigned a UV you are supposed to assign the UV as an input to the image texture node. If a mesh has only one UV, it takes that one as a default and one can assign a UV without explicitly doing it.  When there is more than one UV you can't take the implicit assignment and one needs to be explicit and add a UV node to the material, assign the correct UV and connect it into the image texture.

![image](https://user-images.githubusercontent.com/79840/81162734-90253000-8f85-11ea-8083-a012f32f07fa.png)

I have been working on improving the material generation of the addon to do exactly this.  However if you can't wait, experiment with this feature and see if it helps.  I could use the feedback, to make sure this is the same problem you are seeing and if my solution is the same as your problem.

If not, send on an offending mesh and detail where exactly the problem is

**Comment by douglaskastle** [comment]
Same mesh with out the UV assign:

![image](https://user-images.githubusercontent.com/79840/81163841-45a4b300-8f87-11ea-8080-7126cf45a19a.png)

**Comment by Donald-Drennan** [comment]
Douglas,

That is a problem, but not the one I was concerned with as much.

Take a closer look at these screenshots. In Lightwave, the UVs look right,
but translated to Blender, they're messed up. You see the polygons
stretching between the islands? It's not handling shared points correctly,
and therefore the UVs are unusable. I don't know how to fix this.

Thanks again!

**Comment by douglaskastle** [comment]
There are no images in your post

**Comment by Donald-Drennan** [comment]
That would explain why you didn't understand me. That's odd. I attached
them, it shows that they are attached in my Sent folder.

Try this. Insert vs. attach
[image: Screenshot 2020-05-03 17.02.44.jpg]
[image: Screenshot 2020-05-03 17.04.37.jpg]

Thanks!
Donald L. Drennan


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by Donald-Drennan** [comment]
Douglas,

Sorry! I didn't realize that was happening. I had been replying via email, and github must strip out attachments.

![Screenshot 2020-05-03 17 02 44](https://user-images.githubusercontent.com/64158071/81228849-cf22a800-8fbc-11ea-9f97-41b23da502ac.png)
![Screenshot 2020-05-03 17 04 37](https://user-images.githubusercontent.com/64158071/81228869-d8137980-8fbc-11ea-96c5-f02db9236e52.png)

**Comment by douglaskastle** [comment]
OK I see and understand the problem.  I have no idea what the solution is.

Looking at the LW version, there are red dots on some points, I assume these are co-located points, so they are there effectively twice.  Blender gets confused, renders only one of the 2 and then keeps the join going to the second locations next neighbour.

Ponder!

**Comment by Donald-Drennan** [comment]
Yes, that seems to be the problem. I had lots of trouble with this in
Lightwave when trying to create UV maps back in previous version.
Blender shows co-located points separated. I don't know why it can't deal
with this. It makes the UV map useless.
If you figure out anything, please let me know.

Thanks!

Donald L. Drennan


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by douglaskastle** [comment]
It makes certain types of UV useless.  There are other UV maps that it works with just fine.

I am surprised that this is such a deal breaker for you ;). But I will have a look

**Comment by Donald-Drennan** [comment]
Thanks. Are you saying some Lightwave model's UVs translate correctly? I
will test more of my models.

If this could be a fixed easily in Blender, it wouldn't be a problem, but
getting Blender to create the exact same layout seems imposible, so the
texture mapping doesn't work right. If you try to paint on the bad UVs, you
get slashes of color where it's not supposed to be.

I really appreciate your hard work on this.
Thanks again!

**Comment by douglaskastle** [comment]
Yes if you look at my above images, the two faces with numbers on these are correctly assigned UVs. They are square images on square faces.

Might be a storm in a teacup, I reordered some stuff in the addon and got this:

![image](https://user-images.githubusercontent.com/79840/81237389-6e649100-8ff7-11ea-8dd4-7ea9798990eb.png)

I'll check a few other places to to see if my fix didn't break something else

**Comment by douglaskastle** [comment]
As an FYI, 3 visible UV with 3 separate images:

![image](https://user-images.githubusercontent.com/79840/81237643-0e221f00-8ff8-11ea-8187-1c039948b425.png)

Also of notes, blender can only handle a maximum number of 8 UVs. Not too sure if that is per blend or per mesh.

**Comment by douglaskastle** [comment]
Also I know you commented on this post, but I am recording it here for reference:

https://blenderartists.org/t/uv-unwrapping-baking-woes-and-fundamental-confusion-free-mesh-attached/1222119

**Comment by Donald-Drennan** [comment]
Well that looks like it worked! That's great if nothing else is broken.
I look forward to trying it.

Thanks!
